### DISCUSSION TITLE
Savestate: unrankable mod that sends you back a few seconds every time you miss

### DISCUSSION THREAD

**Discussion by Doksha** [ppy/osu#19932] [Ideas]
Would fall under the same category as relax and autopilot in that it would mainly be targeted towards practicing. Could also be configured to only send you back if you fail (this would probably be much more difficult to implement though).

This would work by keeping the previous inputs (and keep certain "savestates" of score, acc, cursor location, etc.) and would replay the inputs up until the miss, where the game would begin accepting inputs again. (this would solve the problem of shoving you into a place where you dont have the time to hit the circles, and getting in an infinite loop)
A good reference would be Trackmania 2020's checkpoint system, you preserve "momentum" up until the savestate, then regain control.
Maybe have savestates only every n notes (default could be 50 or something) so you cant just retry spam and try to hit the next note every try, (or allow the player to bind a key? Probably a bad idea though.) could also save before diff spikes, though that would likely be difficult to implement.

**Comment by peppy** [comment]
This was posted very recently somewhere else (along with a sample implementation, if I recall). It's also basically practice mod that has been requested for a long time.

**Reply by nagi-desuuu** [reply]
Would this mod be hard to implement code-wise?

**Reply by peppy** [reply]
https://ptb.discord.com/channels/188630481301012481/188630652340404224/1007413720495956141
https://github.com/harshavb/osu/tree/rewind-mod
